File: src/ELibrary.Frontend/src/app/modules/shared/domain/dtos/shop/statistics/bookStatisticsResponse.ts
import { BookStatistics } from "../../../..";

export interface BookStatisticsResponse {
    inCartCopies: number;
    inOrderCopies: number;
    soldCopies: number;
    canceledCopies: number;
    orderAmount: number;
    canceledOrderAmount: number;
    averagePrice: number;
    stockAmount: number;
    earnedMoney: number;
    orderAmountInDays: { date: Date, count: number }[];
}

export function mapBookStatisticsResponseToBookStatistics(response: BookStatisticsResponse): BookStatistics {
    return {
        inCartCopies: response.inCartCopies,
        inOrderCopies: response.inOrderCopies,
        soldCopies: response.soldCopies,
        canceledCopies: response.canceledCopies,
        orderAmount: response.orderAmount,
        canceledOrderAmount: response.canceledOrderAmount,
        averagePrice: response.averagePrice,
        stockAmount: response.stockAmount,
        earnedMoney: response.earnedMoney,
        orderAmountInDays: new Map(
            Object.entries(response.orderAmountInDays).map(
                ([date, count]) => [new Date(date), count as unknown as number]
            )
        )
    }
}

File: src/ELibrary.Backend/ShopApi/Features/StatisticsFeature/Domain/Dtos/BookStatisticsResponse.cs
namespace ShopApi.Features.StatisticsFeature.Domain.Dtos
{
    public class BookStatisticsResponse
    {
        public long InCartCopies { get; set; }
        public long InOrderCopies { get; set; }
        public long SoldCopies { get; set; }
        public long CanceledCopies { get; set; }
        public long OrderAmount { get; set; }
        public long CanceledOrderAmount { get; set; }
        public decimal AveragePrice { get; set; }
        public long StockAmount { get; set; }
        public decimal EarnedMoney { get; set; }
    }
}